std/range/package.d

```diff
@@ -5887,284 +5887,387 @@ pure @safe nothrow unittest
+    else
+        static assert(false, "iota does not support types that cannot be added or incremented.");
+
+    ValueType begin;
+    ValueType end;
+    S step;
+
+    private auto unsigned(T)(T value)
     {
-        private Value current, last;
-        private StepType step; // by convention, 0 if range is empty
+        static if (is(Unsigned!T)) return cast(Unsigned!T) value;
+        else return value;
+    }
 
-        this(Value current, Value pastLast, StepType step)
-        {
-            if (current < pastLast && step > 0)
-            {
-                // Iterating upward
-                assert(unsigned((pastLast - current) / step) <= size_t.max);
-                // Cast below can't fail because current < pastLast
-                this.last = cast(Value) (pastLast - 1);
-                this.last -= unsigned(this.last - current) % step;
-            }
-            else if (current > pastLast && step < 0)
-            {
-                // Iterating downward
-                assert(unsigned((current - pastLast) / (0 - step)) <= size_t.max);
-                // Cast below can't fail because current > pastLast
-                this.last = cast(Value) (pastLast + 1);
-                this.last += unsigned(current - this.last) % (0 - step);
-            }
-            else
-            {
-                // Initialize an empty range
-                this.step = 0;
-                return;
-            }
-            this.step = step;
-            this.current = current;
-        }
+    this(B b, E e, S s)
+    {
+        step = 0;
+        if (s == 0) return;
+        if (b < e && s < 0) return;
+        if (b > e && s > 0) return;
 
-        @property bool empty() const { return step == 0; }
-        @property inout(Value) front() inout { assert(!empty); return current; }
-        void popFront()
+        begin = b;
+        end = e;
+        step = s;
+
+        if (begin < end && step > 0)
         {
-            assert(!empty);
-            if (current == last) step = 0;
-            else current += step;
+            // Iterating upward
+            assert(unsigned((end - begin) / step) <= size_t.max);
+            // Cast below can't fail because begin < end
+            end = cast(E) (end - 1);
+            end -= unsigned(end - begin) % step;
         }
-
-        @property inout(Value) back() inout
+        else if (begin > end && step < 0)
         {
-            assert(!empty);
-            return last;
+            // Iterating downward
+            assert(unsigned((begin - end) / (0 - step)) <= size_t.max);
+            // Cast below can't fail because begin > end
+            end = cast(E) (end + 1);
+            end += unsigned(begin - end) % (0 - step);
         }
-        void popBack()
+        else
         {
-            assert(!empty);
-            if (current == last) step = 0;
-            else last -= step;
+            // Initialize an empty range
+            step = 0;
```

Comment:
This seems overdone. There have been several tests above that eliminated the empty range cases. There would be the case where `begin == end` and `step != 0` but those should be handled properly as limits of the normal cases.